From patchwork Fri Dec 14 22:34:43 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10731769
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 0AA0413AD
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 14 Dec 2018 22:35:01 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id AB2C22D787
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 14 Dec 2018 22:35:00 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 9CF362D78F; Fri, 14 Dec 2018 22:35:00 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 2C2852D787
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 14 Dec 2018 22:35:00 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1729707AbeLNWe6 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 14 Dec 2018 17:34:58 -0500
Received: from mail-vs1-f74.google.com ([209.85.217.74]:47909 "EHLO
        mail-vs1-f74.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1729341AbeLNWe6 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 14 Dec 2018 17:34:58 -0500
Received: by mail-vs1-f74.google.com with SMTP id y139so3224210vsc.14
        for <kvm@vger.kernel.org>; Fri, 14 Dec 2018 14:34:57 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=YPH6TPsgMa/4T66cYZq3Q+uaX57wZSk4vAaYPUfD3Mc=;
        b=AL97G6m0syEAq5O117TpvaUdLv9SaHuI299ZEMIZL/QULctReItzVkdTilTiRaAUCY
         zVZxRnT23pC0ETYvdSxl/RdLPpuJNSWF6SuDoDNDeo6semx/9u4p5ovf6ij6pFOg59RG
         dbVqtTrs8Z4nUf38C46hMAVB87n4LyjyogiaxflojAoEfpEODb3JhVqjidBqnyndH8uM
         cDZcUGgjS1oXPLIsDRMTIgiCYDlznro65lCQ6wxJ/IuNLk9Wj1TsZXqHlakG/Tq+LpPd
         fzbAbEv8rmMTRVuuc27Po2mXfqstrdTvZZw8sje5yj7w10C94tGZdkAZ87e7vMO0z6B6
         1aHw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=YPH6TPsgMa/4T66cYZq3Q+uaX57wZSk4vAaYPUfD3Mc=;
        b=dSKuysD3GwwQAczQG44pD3T+8or656BBPF9xQs+20J0vbflWckzDmcU6vhnokzv+YX
         4HBFORCDYonrQZDgBevIboCZeCfjr3P4vujEIkB/6HmO7eXXQEp11buCCpqxtPXZ9tmr
         0MiECnJOeMEtEp3K35FUuXNjTplZLgTovP8Fxh8fIWtHYCna9t0qq44h965BqFtSn/WH
         9G3RoI8HH9sNtEe5zRgiL7qPlSVcojXgR4nfyO4WpPneWGNR2LjbPpdxNnfRtKwVW1Ni
         zB6anfve8QDD5EW7B/mSb+NplKh0Uj//z0VaUx7g4puCwPH8PHznUABr32C4eFcC99rS
         IvZw==
X-Gm-Message-State: AA+aEWYq6e9kLnzxufoB3zMDDWn/pdQIRdZdwyes/kDVHzqDpRnayBTf
        Sc6m7Ke+MaXtyFHfrI2jinmCaBvuKg25pPc26wY298TCpdRjJmX1omyr7P3C0pHANdQpiZKkOIC
        yA9J9ANdXedmuB+VtytKjQIgNUXdPl0EjXbhzqFCSWrYbCFjqmgdagfK8jJ8ew4U=
X-Google-Smtp-Source: 
 AFSGD/XnVBuZfbrt0SUNlXuZ7gkvdysgSJjM/IBs78KuKgp4ZTsH/Bw/jLpI52mGYX/LISsFmO/ss1oO48K36g==
X-Received: by 2002:a1f:abcf:: with SMTP id u198mr4646337vke.21.1544826897168;
 Fri, 14 Dec 2018 14:34:57 -0800 (PST)
Date: Fri, 14 Dec 2018 14:34:43 -0800
In-Reply-To: <20181214203239.GH22063@linux.intel.com>
Message-Id: <20181214223443.153614-1-jmattson@google.com>
Mime-Version: 1.0
References: <20181214203239.GH22063@linux.intel.com>
X-Mailer: git-send-email 2.20.0.405.gbc1bbc6f85-goog
Subject: [PATCH v2] kvm: Change offset in kvm_write_guest_offset_cached to
 unsigned
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>, Cfir Cohen <cfir@google.com>,
        Peter Shier <pshier@google.com>,
        Krish Sadhukhan <krish.sadhukhan@oracle.com>,
        Sean Christopherson <sean.j.christopherson@intel.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Since the offset is added directly to the hva from the
gfn_to_hva_cache, a negative offset could result in an out of bounds
write. The existing BUG_ON only checks for addresses beyond the end of
the gfn_to_hva_cache, not for addresses before the start of the
gfn_to_hva_cache.

Note that all current call sites have non-negative offsets.

Fixes: 4ec6e8636256 ("kvm: Introduce kvm_write_guest_offset_cached()")
Reported-by: Cfir Cohen <cfir@google.com>
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Cfir Cohen <cfir@google.com>
Reviewed-by: Peter Shier <pshier@google.com>
Reviewed-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
Reviewed-by: Sean Christopherson <sean.j.christopherson@intel.com>
---
 include/linux/kvm_host.h | 3 ++-
 virt/kvm/kvm_main.c      | 3 ++-
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/include/linux/kvm_host.h b/include/linux/kvm_host.h
index c926698040e0d..a03d5e264e5e7 100644
--- a/include/linux/kvm_host.h
+++ b/include/linux/kvm_host.h
@@ -694,7 +694,8 @@ int kvm_write_guest(struct kvm *kvm, gpa_t gpa, const void *data,
 int kvm_write_guest_cached(struct kvm *kvm, struct gfn_to_hva_cache *ghc,
 			   void *data, unsigned long len);
 int kvm_write_guest_offset_cached(struct kvm *kvm, struct gfn_to_hva_cache *ghc,
-			   void *data, int offset, unsigned long len);
+				  void *data, unsigned int offset,
+				  unsigned long len);
 int kvm_gfn_to_hva_cache_init(struct kvm *kvm, struct gfn_to_hva_cache *ghc,
 			      gpa_t gpa, unsigned long len);
 int kvm_clear_guest_page(struct kvm *kvm, gfn_t gfn, int offset, int len);
diff --git a/virt/kvm/kvm_main.c b/virt/kvm/kvm_main.c
index 2679e476b6c39..065ee2fb40340 100644
--- a/virt/kvm/kvm_main.c
+++ b/virt/kvm/kvm_main.c
@@ -1965,7 +1965,8 @@ int kvm_gfn_to_hva_cache_init(struct kvm *kvm, struct gfn_to_hva_cache *ghc,
 EXPORT_SYMBOL_GPL(kvm_gfn_to_hva_cache_init);
 
 int kvm_write_guest_offset_cached(struct kvm *kvm, struct gfn_to_hva_cache *ghc,
-			   void *data, int offset, unsigned long len)
+				  void *data, unsigned int offset,
+				  unsigned long len)
 {
 	struct kvm_memslots *slots = kvm_memslots(kvm);
 	int r;
